homeassistant/components/sma/__init__.py

```diff
@@ -1 +1,193 @@
+            + list(pysma.LEGACY_MAP)
+        )
+    )
+
+    for sensor in possible_sensors:
+        if sensor in sensor_def:
+            pysma_sensor = sensor_def[sensor]
+            original_key = pysma_sensor.key
+        elif sensor in pysma.LEGACY_MAP:
+            # If sensor was removed from pysma we will remap it to the new sensor
+            legacy_sensor = pysma.LEGACY_MAP[sensor]
+            pysma_sensor = sensor_def[legacy_sensor["new_sensor"]]
+            original_key = legacy_sensor["old_key"]
+        else:
+            _LOGGER.error("%s does not exist", sensor)
+            continue
+
+        # Find entity_id using previous format of unique ID
+        entity_id = entity_registry.async_get_entity_id(
+            "sensor", "sma", f"sma-{original_key}-{sensor}"
+        )
+
+        if entity_id:
+            # Change entity_id to new format using the device serial in entry.unique_id
+            new_unique_id = (
+                f"{entry.unique_id}-{pysma_sensor.key}_{pysma_sensor.key_idx}"
+            )
+            entity_registry.async_update_entity(entity_id, new_unique_id=new_unique_id)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up sma from a config entry."""
+    # Init all default sensors
+    sensor_def = pysma.Sensors()
+
+    if entry.source == SOURCE_IMPORT:
+        await _parse_legacy_options(entry, sensor_def)
+        await _migrate_old_unique_ids(hass, entry, sensor_def)
+
+    # Init the SMA interface
+    protocol = "https" if entry.data.get(CONF_SSL) else "http"
+    url = f"{protocol}://{entry.data.get(CONF_HOST)}"
+    verify_ssl = entry.data.get(CONF_VERIFY_SSL)
+    group = entry.data.get(CONF_GROUP)
+    password = entry.data.get(CONF_PASSWORD)
+
+    session = async_get_clientsession(hass, verify_ssl=verify_ssl)
+    sma = pysma.SMA(session, url, password, group)
+
+    # Ensure we logout on shutdown
+    async def async_close_session(event):
+        """Close the session."""
+        await sma.close_session()
+
+    remove_stop_listener = hass.bus.async_listen_once(
+        EVENT_HOMEASSISTANT_STOP, async_close_session
+    )
+
+    # Define the coordinator
+    async def async_update_data():
+        """Update the used SMA sensors."""
+        values = await sma.read(sensor_def)
+        if not values:
+            raise UpdateFailed
+
+    interval = entry.options.get(CONF_SCAN_INTERVAL) or timedelta(
+        seconds=DEFAULT_SCAN_INTERVAL
+    )
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name="sma",
+        update_method=async_update_data,
+        update_interval=interval,
+    )
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = {
+        PYSMA_OBJECT: sma,
+        PYSMA_COORDINATOR: coordinator,
+        PYSMA_SENSORS: sensor_def,
+        PYSMA_REMOVE_LISTENER: remove_stop_listener,
+    }
+
+    await coordinator.async_config_entry_first_refresh()
```

Comment:
```suggestion
    await coordinator.async_config_entry_first_refresh()

    hass.data[DOMAIN][entry.entry_id] = {
        PYSMA_OBJECT: sma,
        PYSMA_COORDINATOR: coordinator,
        PYSMA_SENSORS: sensor_def,
        PYSMA_REMOVE_LISTENER: remove_stop_listener,
    }
```

If setup fails and `async_config_entry_first_refresh` raises `ConfigEntryNotReady`, this will result in a memory leak of a listener each time since it will never be canceled.

```
    remove_stop_listener = hass.bus.async_listen_once(
        EVENT_HOMEASSISTANT_STOP, async_close_session
    )
```

Possible solutions would be to install the listener after the first refresh, if we assume the session doesn't need to be closed if the first refresh fails.

or you could something like

```python
   try: 
      await coordinator.async_config_entry_first_refresh()
   except ConfigEntryNotReady:
      await sma.close_session()
      remove_stop_listener()
      raise
```

Reply:
Did a bit of both. Moved the listener install and creation of `hass.data[DOMAIN][entry.entry_id]` after the first refresh and call `sma.close_session()` if the refresh raises `ConfigEntryNotReady`

https://github.com/home-assistant/core/commit/2ad7ad8c9aae7286e00dd578cbf2342eac6bcbc4#diff-a7bdb8a0a15589eb3d6f0a9b41a9131ef4a2563166a67a1536b973a7534adcfeL132-L169